# Análisis exploratorio de datos

En este notebook se trabajará con el dataset **Titanic** para poder aplicar las medidas vistas en clase  a un dataset

## Objetivos
- reconocer variables numéricas y categóricas
- calcular medidas de tendencia central
- calcular medidas de dispersión
- construir tablas de frecuencias
- calcular el índice de Gini de heterogeneidad
- Dar interpretaciones a los resultados obtenidos



## 1. Bibliotecas
Bibliotecas que se utilizaràn:
- `numpy` para cálculo numérico
- `pandas` para manejo de datos
- `matplotlib` para gráficas
- `seaborn` para cargar el dataset de Titanic y crear gráficas para la visualización


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import seaborn as sns

## 2. Carga del dataset

Carga el dataset que se encuentra en `seaborn`


In [2]:
df = sns.load_dataset("titanic")
df

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


## 3. Exploración de los datos

### Actividad 1
Muestra:
1. el número de filas y columnas
2. los nombres de las columnas
3. las primeras 5 filas
4. Escribe que significa cada variable, el tipo que es (cualitativa, cuantitativa etc)

Usar `df.shape`, `df.columns` y `df.head()` `df.columns`


In [3]:
# numero de filas y columnas
filas, columnas = df.shape
print("Filas:", filas)
print("Columnas:", columnas)

# nombres de las columnas
print(list(df.columns))

# primeras 5 filas
df.head()

Filas: 891
Columnas: 15
['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town', 'alive', 'alone']


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


| VARIABLE | SIGNIFICADO | TIPO |
|---|---|---|
| survived | sobrevivió? 0 = no, 1 = sí | cualitativa nominal |
| pclass | Passenger's ticket class. 1st=Upper, 2nd=Middle, 3rd=Lower | cualitativa ordinal |
| sex | Sexo biológico | cualitativa nominal |
| age | Edad al momento del incidente | cuantitativa discreta |
| sibsp | número de hermanos - esposos abordo | cuantitativa discreta |
| parch | número de padres - hijos abordo | cuantitativa discreta |
| fare | Tarifa | cuantitativa continua |
| embarked | Puerto donde embarcó. C = Cherbourg, Q = Queenstown, S = Southampton | cualitativa nominal |
| class | clase de pasajero. First = primera, Third = tercera | cualitativa ordinal |
| who | si es hombre o mujer | cualitativa nominal |
| adult_male | si es un hombre (sexo) adulto | cualitativa nominal |
| deck | Indica la cubierta en la que se hospedaba el pasajero | cualitativa ordinal |
| embark_town | poblado donde embarcó | cualitativa nominal |
| alive | sigue vivo? | cualitativa nominal |
| alone | iba solo? | cualitativa nominal |

## 4. Resumen del dataset

### Actividad 3
Explora la estructura general del dataset

Usa `df.info()` y `df.describe()`


Con `df.info()` podemos observar los datos del dataset:



*   RangeIndex nos indica que hay 891 filas enumeradas de 0 a 890
*   Indica el total de columnas, en este caso 15
*   Tenemos una estructura con la información de cada columna, por cada una de estas nos indica el número que le corresponde (de 0 a 14), el nombre de la columna, el conteo de cuantas filas son non-null en la columna correspondiente y el tipo de dato que almacena la columna.

Así también indica el número de columnas por cada tipo de dato, es decir, bool(2) nos dice que hay dos columnas con tipo de dato bool.




In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


Con `df.describe()` observamos las estadísticas del dataset.

Como columnas tenemos las correspondientes a la tabla pero únicamente las que tienen tipo de dato numérico int o float, no observamos otros tipos.

Por cada columna calcula:



*   *count*: Cuantas filas contiene la columna
*   *mean*: Media
*   *std*: Desviación estándra
*   *min*: Valor mínimo
*   *25%*: Primer cuartil (Q1)
*   *50%*: Mediana (Q2)
*   *75%*: Tercer cuartil (Q3)
*   *max*: Valor máximo



In [5]:
df.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### Actividad 4
De acuerdo a la salida anterior y responde:

1. ¿Qué variables tienen valores nulos?
2. ¿Qué variable elegirías empezar a calcular medidas de tendencia central?
3. ¿Qué variable elegirías para trabajar frecuencias?



**Respuesta:**  
1.  
2.  
3.  


## 5. Medidas de tendencia central

Se tiene una variable **`fare`**, lo cuál indica cuánto pago un pasajero


### Actividad 5
Guarda la variable `fare` sin valores nulos en una nueva variable.

Usa `dropna()`.


### Actividad 6
Calcula:
- media de `fare`
- mediana de `fare`
- moda de `fare`

Usa `.mean()`, `.median()` y `.mode()`.


### Actividad 7
Interpreta tus resultados.

1. ¿La media y la mediana son parecidas o diferentes?
2. ¿Qué podría indicar esa diferencia?
3. ¿Qué representa la moda en este caso?

**Respuesta:**  
1.  
2.  
3.  


## 6. Medidas de dispersión

**¿los valores están muy juntos o muy separados?**


### Actividad 8
Calcula para `fare`:
- mínimo
- máximo
- rango manual (`max - min`)
- rango con `np.ptp`
- varianza
- desviación estándar

Usar `.min()`, `.max()`, `.var()`, `.std()` y `np.ptp(...)`.


### Actividad 9
Interpretación:

1. ¿La variable `fare` tiene mucha o poca variabilidad?
2. ¿Qué indica el rango?
3. ¿Qué interpretación le darías a la desviación estándar?



**Respuestas:**  
1.  
2.  
3.  


## 7. Visualización

### Actividad 10
Construye un histograma de `fare`.

Usar `plt.hist(...)`.


### Actividad 11
Construye un boxplot de `fare`.

Usar `plt.boxplot(...)`.


### Actividad 12
De acuerdo a la gráfica anterior y a las medidas calculadas previamente, responde:

- ¿Se observan valores atípicos?
- ¿Eso ayuda a explicar la diferencia entre media y mediana?
- ¿Qué gráfico te ayudó más a entender la dispersión?


**Tu interpretación:**  


## 8. Frecuencias de una variable categórica

Se usará la columna **`class`**.

Las frecuencias sirven para responder: **¿cuántos casos hay en cada categoría?**


### Actividad 13
Calcula la frecuencia absoluta de `class`.

Usa `value_counts()`.


### Actividad 14
Calcula la frecuencia relativa de `class`.

 Puedes dividir las frecuencias absolutas entre el total o usar `df["class"].value_counts(normalize=True)`


### Actividad 15
Interpreta la distribución de `class`.

- ¿La variable está equilibrada?
- ¿Hay una categoría dominante?
- ¿Qué información útil puedes obtener de esa tabla?

**Respuestas:**  


## 9. Índice de Gini de heterogeneidad

Este índice se usa con variables categóricas para medir qué tan repartidas están las categorías.



### Actividad 16
Exploa la columna `class`:

1. Realiza el cálculo del índice Gini para la variable class




### Actividad 17
Interpreta el resultado del índice de Gini.

- ¿El valor obtenido sugiere mucha o poca heterogeneidad y por qué?
- ¿Qué pasaría si una sola categoría concentrara casi todos los casos?


**Respuestas:**  


## 10. Resumen

### Actividad 18
Responde con tus propias palabras:

- La media y la mediana sirven para...
- La desviación estándar sirve para...
- Las frecuencias sirven para...
- El índice de Gini de heterogeneidad sirve para...



## 12. Opcional

Utiliza la variable  `survived` u otra que sea de tu interés para calcular:
- tabla de frecuencias
- proporciones
- índice de Gini de heterogeneidad
- Escribe tu interpretación


## Bibliografía

1. https://numpy.org/doc/
2. https://pandas.pydata.org/
3. https://seaborn.pydata.org/
4. https://github.com/mwaskom/seaborn-data/blob/master/titanic.csv